### DISCUSSION TITLE
Allow nested layers within a slice

### DISCUSSION THREAD

**Discussion by H1op3ts** [feature-sliced/documentation#716] [Ideas]
The main idea is allowing the same layers structure but within a slice. If some slice (let it be some page for example) has it's own **UNIQ** business logic we can use the same **FSD** approach to structurize that logic within a slice folder i.e we can place "features" and "widgets" folders alogside with model/api/ui within that page slice. The rules are exactly the same as for root layers: only one direction relations. Layers within a slice could **ONLY** start with the layer below the slice layer, i.e if a slice is an instance of pages it's own layers could start with widgets, if a slice is instance of widgets then it's own layers could start from features, etc. This will allow to limit nesting structure by layers depth and nesting will be finite. There is no violation in case if some nested feature imports some "global feature layer" feature, because nested feature relates to it's higher slice, there couldn't be cross imports issues. Possibly it's better to place nested layers within a "model" folder.

**Nested structure example**

```
-app
|-pages
|---page-slice
|---|--model
|---|--ui
|---|--lib
|---|--api
|---|--widgets
|---|--|--page-slice-uniq-widget
|---|--|--|--model
|---|--|--|--ui
|---|--|--|--features
|---|--|--|--|--page-slice-uniq-widget-feature
|---|--|--|--|--|--model
|---|--|--|--|--|--ui
|---|--features
|---|--|--page-slice-uniq-feature
|---|--|--|--model
|---|--|--|--ui
 ...
|-widgets
|---widget-slice
|---|--model
|---|--ui
|---|--lib
|---|--api
|---|--features
|---|--|--widget-slice-uniq-feature
|---|--|--|--model
|---|--|--|--ui
|-features
  ...
|-entities
  ...
```

**Comment by illright** [comment]
There has been some talk in the core-team about fractal FSD structure, but none of it has been released yet.

As for your proposal, my main question is: why do we need this? Layer names have semantic meaning, so when we consider the term "FSD Widget", we expect to find all of them in `src/widgets`, but with this proposal we would also have to check all higher layers for other widgets. Of course, we might consider distinguishing between global widgets and local widgets, but then why do we even need to call the local widgets _widgets_? I think it can just as easily be a plain component.

I think that the main struggle that this proposal is trying to address is how to organize a lot of code within a single slice. It's a very valid struggle, but addressing it properly would probably require some examples of complex logic inside a slice when a simple division by segments is no longer enough. Could you come up with some example that illustrates the benefits of this proposal over the existing structure?

**Reply by H1op3ts** [reply]
TBH I came to this proposal from the implementation based on the existing approach.

**Case 1:**
Some higher slices could have very comprehensice business logic behind them, although this logic could be possibly and spread across the "global" layers some of the pieces of that logic is very specific to a slice(to a page in this particular example) therefore we'd like to keep this pieces within a page slice. The issue is that even these slice-specific pieces represent a very complex logic so we'd like to orginize them as well. To be more specific we have a page with some kind of navigator. That navigator component is compatible to this specific page only it couldn't be reused elsewhere. Although it might looks quite simple in terms of the UI it contains huge amount of business logic so we'd like to split it into features as well so in terms of semantic definition it's still a very suitable instance of widget.

**Case 2:**
Slice-specific nested features could be a special case of the more abstract "global" feature. Keeping them within a higher slice allows to imprort "global" feature without any violations and easily build slice specific feature on top of the core feature. Imagine that we have some abstract "global" search feature (not a regular one but something more complex, like advanced search with more fields and conditions). For some page you need this search to be more specific: working with some certain entities, have additional conditions and inputs, etc. You organize page slice to have it's own advanced search feature and keep it under semantic layer "feature" in order to keep consistency with core structure.

Case 1 reference:
<img width="256" alt="Screenshot 2024-09-01 at 13 54 06" src="https://github.com/user-attachments/assets/7d88081f-6f9c-499e-aacb-0e53d62ceb00">

**Reply by illright** [reply]
Could you collect some feedback on this case in our Telegram chat? So far I'm not convinced that this case is not sufficiently covered by the current specification of FSD

**Reply by sanua356** [reply]
@H1op3ts Hi. The proposal is interesting, but I think it will make the architecture more difficult to understand and more confusing. At the moment, I don't think it's worth resorting to this kind of recursive layering.

@illright Hi, Could you please take a look at my open Pull Request #711 in the documentation repository?

**Reply by H1op3ts** [reply]
@sanua356 The proposal is most likely not useful for relatively small projects but could be really  helpful for heavy monolyths. It should not complicate the architecture, precisely because it is recursive (but still finite) and uses a single set of rules. In other words your higher slices (could) became a micro-project (which keeps all its own uniq logic to itself) with the same FSD strucutre

**Reply by H1op3ts** [reply]
@illright Got ya, perhaps we'll get a try on our project and will get back with some measurable feedback

**Reply by sanua356** [reply]
@H1op3ts I have the exact opposite opinion regarding this proposal. For small projects, in my opinion, you can try to use it, but for large monoliths, it seems that such a structure will only confuse developers more

**Reply by H1op3ts** [reply]
@sanua356 what kind of architecture you'd recommend then? Let's say for 2 cases I've described above

**Reply by sanua356** [reply]
@H1op3ts In both cases, I don't see a problem in decomposing the logic to such a level that the layers correspond to the current FSD specification.

**Reply by H1op3ts** [reply]
@sanua356 Do you mean move everything to global "widgets" and "layers" or spread everything accross model/ui/lib within a page slice?

**Reply by sanua356** [reply]
@H1op3ts Distribute the necessary logic within the layer (model/ui/lib)

**Reply by H1op3ts** [reply]
@sanua356 Yep, got it. That's actually a problem. Since the logic it relatively complex you still have to stick to some rules to organize it. And the question is why do I need to invent different set of rules when I can use exactly the same one, from the global level

**Reply by sanua356** [reply]
@H1op3ts Of course you can stick to the established set of rules in your case, but I don't think it is necessary to make it the FSD standard.

**Reply by H1op3ts** [reply]
@sanua356 Can't agree more, I don't insist on that to be a standard, much less  on that to be mandatory. Just to officially ALLOW use such a nesting

**Reply by illright** [reply]
Well, FSD is extendable, so there's nothing stopping you from implementing this in your projects, but then you would be responsible for maintaining the documentation for this extension. The reason I'm hesitant to pull it into the main documentation is because it might complicate the learning process even more than it is already.

**Reply by sanua356** [reply]
@illright Yes. I agree. Could you please review pull request #711?

**Reply by illright** [reply]
@sanua356 I will review that pull request when I have the time and energy. I remember about it, no need to remind about it further.

**Comment by noveogroup-amorgunov** [comment]
Hi! I’d like to bring this topic back up, as the idea has become especially relevant after the release of version 2.1.

Page slices are becoming quite large and starting to contain a significant amount of code. Let’s take https://samokat.ru/ as an example. On the homepage, there are various sliders with products, sliders with promo selections and recipes, category listings, and informational banners. Overall, this is similar to any admin dashboard.

If we place all this in page segments, we end up with so-called destructive decoupling, where modules are split across different segments of the page.

```
src
  pages
    home
      index.ts
      ui
        MainPageContainer/
        AwesomeProductSlider/
        AwesomeProductSliderCard/
        RecipeSlider/
        RecipeSliderCart/
        CategoryListing/
        CategoryCard/
        ShowcaseInformersGrid/
        ... // a lot of other components
      model
        awesomeProductSlice.ts
        pageSlice.ts
        types.ts
      api
        awesomeProductApi.ts
        showcaseInformerApi.ts
        categoryTreeApi.ts
        ...
      lib
        ...apiMappers
```

For instance, the "awesome products" slider ends up scattered across page segments. It would be great to group it into a standalone slice (like a widget). Moving it to the global widgets layer is not an option, since it isn’t used anywhere else yet.

If sub-slices (similar to fractals) were allowed, the project structure could look something like this:

```diff
src
  pages
    home
      index.ts
+      @local-entities
+         recipe/
+      @local-widgets
+        awesome-product-slider/
+          ui/
+            AwesomeProductSlider/
+            AwesomeProductSliderCard/
+          model/
+            awesomeProductSlice.ts
+            types.ts
+          api
+            awesomeProductApi.ts
+          lib
+            mapAwesomeProductsDto.ts
+        recipe-slider/
+        category-listing/
+        showcase-informers-grid/
+        ...
      ui
        // only base components for page
        MainPageContainer/
      model
        pageSlice.ts
        types.ts
```

Довран wrote about introducing fractals in https://t.me/feature_sliced/1/136388. In my opinion, it’s a great idea that doesn’t break current compatibility and adds a lot of flexibility - **sub-slices**. Additionally, it would help promote the page-first approach, where we build applications starting from the page layer.

Довран also suggested limiting the nesting to one level, which I agree with. Exactly how to structure fractals doesn’t seem that critical. Let it be an additional segment, with a prefix like @... to explicitly distinguish it in the structure.

> With fractals, I think the blocker lies elsewhere—what are the rules for creating them? For example, is it appropriate to have a pages or app layer within a fractal inside pages/X/?

In response to the question above, I’d suggest not enforcing strict rules, but as a recommendation, advise against using fractals of layers above the current one (for example, allow local widgets within a widget, but discourage local pages inside a widget).

This feature should be documented separately (just like @x) in an advanced section of the documentation.

**Reply by noveogroup-amorgunov** [reply]
I prepare example with this approach https://github.com/noveogroup-amorgunov/fsd-lessons/tree/main/packages/custom-fractal-sub-slices

**Comment by sanctionic** [comment]
@noveogroup-amorgunov In our team we eventually agreed on moving the slides (in your exaple it's "awesome products" widget) into  root fsd/wigdets but under a specific folder like **fds/widgets/@home** that holds everything related to the "home" page. So in general it still in the root widgets slice even though it's not used anywhere except for "home". I personoally don't like it, but I decided to not go against the team:)

**Reply by sanctionic** [reply]
BTW @H1op3ts is my previous account

**Reply by noveogroup-amorgunov** [reply]
@sanctionic The team is sacred, so you made the right choice :) But in my opinion, it’s not very scalable for two reasons:

- In addition to widgets for pages, we might also need sub-widgets for widgets, or features for pages/widgets. And the `@home` notation only covers one use case (widgets for pages).
- If we have many pages (say 50 or more) and they are grouped (grouped slices), we’ll also need to maintain this structure separately on the widgets layer.

But overall, I completely agree that sub-slices give a big boost to expanding the methodology.

**Reply by sanctionic** [reply]
@noveogroup-amorgunov yeah that's exactly what I don't like in our current approach, however as @illright already mentioned - "Well, FSD is extendable, so there's nothing stopping you from implementing this in your projects"
